File: js/packages/browser/src/model_answer.js
/**
 * Deterministic model-answer judge for Browser 1.20.
 *
 * Scores a live provider answer transcript against the pages that were
 * actually fetched. No model, no network, no fallback: every gate is a
 * string/URL/set check and anything unverifiable fails closed.
 *
 * Snake_case wire format mirrors the Python implementation exactly.
 */

const URL_PATTERN = /https?:\/\/[^\s)"'\]]+/g;

function normalize(text) {
  return String(text ?? "").toLowerCase().replace(/\s+/g, " ").trim();
}

function stripTrailingPunctuation(url) {
  return String(url).replace(/[.,;:!?)\]]+$/u, "");
}

function answerUrls(answer) {
  const found = new Set();
  for (const match of String(answer ?? "").matchAll(URL_PATTERN)) {
    found.add(stripTrailingPunctuation(match[0]));
  }
  return found;
}

function asList(value) {
  return Array.isArray(value) ? value : [];
}

function tokens(text) {
  return normalize(text).replace(/[^a-z0-9\s]/g, " ").split(" ").filter(Boolean);
}

function tokenOverlap(quote, markdown) {
  const quoteTokens = new Set(tokens(quote));
  if (!quoteTokens.size) return 0;
  const pageTokens = new Set(tokens(markdown));
  let hits = 0;
  for (const token of quoteTokens) if (pageTokens.has(token)) hits += 1;
  return hits / quoteTokens.size;
}

/**
 * @param {object} transcript {question_id, question, model, answer, claims, citations, pages}
 * claim: {claim_id, statement, citation_urls}; citation: {url, quote}; page: {url, markdown}
 * @param {object} opts {mode: "literal"|"fuzzy", minOverlap}
 */
export function judgeModelAnswer(transcript, opts = {}) {
  const doc = transcript && typeof transcript === "object" ? transcript : {};
  const mode = opts?.mode === "fuzzy" ? "fuzzy" : "literal";
  const minOverlap = Number(opts?.minOverlap ?? 0.6);
  const pages = new Map();
  for (const page of asList(doc.pages)) {
    if (page && typeof page.url === "string") pages.set(page.url, String(page.markdown ?? ""));
  }
  const citations = asList(doc.citations);
  const claims = asList(doc.claims);
  const gates = [];
  const push = (id, name, result, detail = "") => gates.push({ id, name, result, detail });

  const answer = String(doc.answer ?? "");
  push(
    "answer_present",
    "answer is non-empty",
    answer.trim() ? "pass" : "fail",
    answer.trim() ? "" : "empty answer text",
  );

  const citationVerdicts = citations.map((citation, index) => {
    const url = typeof citation?.url === "string" ? citation.url : "";
    const resolves = pages.has(url);
    const quoteText = normalize(citation?.quote ?? "");
    let quoteOk = resolves && quoteText !== "" && normalize(pages.get(url)).includes(quoteText);
    let overlap = quoteOk ? 1 : 0;
    if (!quoteOk && resolves && mode === "fuzzy" && quoteText !== "") {
      overlap = tokenOverlap(citation.quote, pages.get(url));
      quoteOk = overlap >= minOverlap;
    }
    return { index, url, resolves, quoteOk, overlap };
  });
  const unresolved = citationVerdicts.filter((item) => !item.resolves);
  push(
    "citations_resolve",
    "every citation url was fetched",
    citations.length === 0
      ? (claims.length === 0 ? "pass" : "fail")
      : (unresolved.length === 0 ? "pass" : "fail"),
    unresolved.length === 0 ? "" : `unfetched: ${unresolved.map((item) => item.url).join(", ")}`,
  );

  const badQuotes = citationVerdicts.filter((item) => !item.quoteOk);
  const worstOverlap = badQuotes.length
    ? Math.max(...badQuotes.map((item) => item.overlap ?? 0))
    : 1;
  push(
    "quotes_literal",
    mode === "fuzzy" ? "every citation quote overlaps its page" : "every citation quote matches its page literally",
    badQuotes.length === 0 ? "pass" : "fail",
    badQuotes.length === 0
      ? ""
      : `mismatch at citation index ${badQuotes.map((item) => item.index).join(", ")} (mode=${mode}, overlap=${worstOverlap.toFixed(2)})`,
  );

  const uncovered = [];
  for (const claim of claims) {
    const urls = asList(claim?.citation_urls).filter((url) => typeof url === "string" && url !== "");
    if (urls.length === 0 || urls.some((url) => !pages.has(url))) {
      uncovered.push(claim?.claim_id ?? "?");
    }
  }
  push(
    "claims_covered",
    "every claim points at fetched evidence",
    uncovered.length === 0 ? "pass" : "fail",
    uncovered.length === 0 ? "" : `uncovered: ${uncovered.join(", ")}`,
  );

  const invented = [...answerUrls(answer)].filter((url) => !pages.has(url));
  push(
    "no_invented_urls",
    "answer links only fetched pages",
    invented.length === 0 ? "pass" : "fail",
    invented.length === 0 ? "" : `invented: ${invented.join(", ")}`,
  );

  const passed = gates.filter((gate) => gate.result === "pass").length;
  return {
    format: "handoffkit.browser.model_answer_judgment",
    format_version: 1,
    question_id: doc.question_id ?? "",
    model: doc.model ?? "",
    mode,
    gates,
    score: gates.length === 0 ? 0 : passed / gates.length,
    verdict: passed === gates.length ? "pass" : "fail",
    notice: "Deterministic judgment only; no claim beyond the fetched pages.",
  };
}

File: python/packages/handoffkit/handoffkit/browser/model_answer.py
"""Deterministic model-answer judge for Browser 1.20.

Scores a live provider answer transcript against the pages that were
actually fetched. No model, no network, no fallback: every gate is a
string/URL/set check and anything unverifiable fails closed.

Snake_case wire format mirrors the JavaScript implementation exactly.
"""

from __future__ import annotations

import re
from typing import Any

_URL_RE = re.compile(r"https?://[^\s)\"'\]]+")
_TRAILING_PUNCT = ".,;:!?)]"


def _norm(text: Any) -> str:
    return " ".join(str(text or "").lower().split())


def _tokens(text: Any) -> list[str]:
    return [t for t in re.sub(r"[^a-z0-9\s]", " ", _norm(text)).split(" ") if t]


def _token_overlap(quote: Any, markdown: Any) -> float:
    quote_tokens = set(_tokens(quote))
    if not quote_tokens:
        return 0.0
    page_tokens = set(_tokens(markdown))
    return sum(1 for token in quote_tokens if token in page_tokens) / len(quote_tokens)


def _strip_trailing(url: str) -> str:
    return str(url).rstrip(_TRAILING_PUNCT)


def _answer_urls(answer: Any) -> set[str]:
    return {_strip_trailing(m) for m in _URL_RE.findall(str(answer or ""))}


def _as_list(value: Any) -> list[Any]:
    return list(value) if isinstance(value, list) else []


def judge_model_answer(transcript: Any, opts: dict[str, Any] | None = None) -> dict[str, Any]:
    """Judge one model-answer transcript. Fail-closed on any unverifiable part.

    opts: {"mode": "literal"|"fuzzy", "min_overlap": float}. Fuzzy mode passes
    quotes whose token overlap with the page reaches min_overlap (default 0.6).
    """
    doc = transcript if isinstance(transcript, dict) else {}
    options = opts if isinstance(opts, dict) else {}
    mode = "fuzzy" if options.get("mode") == "fuzzy" else "literal"
    try:
        min_overlap = float(options.get("min_overlap", 0.6))
    except (TypeError, ValueError):
        min_overlap = 0.6
    pages: dict[str, str] = {}
    for page in _as_list(doc.get("pages")):
        if isinstance(page, dict) and isinstance(page.get("url"), str):
            pages[page["url"]] = str(page.get("markdown") or "")
    citations = _as_list(doc.get("citations"))
    claims = _as_list(doc.get("claims"))
    gates: list[dict[str, Any]] = []

    def push(gate_id: str, name: str, result: str, detail: str = "") -> None:
        gates.append({"id": gate_id, "name": name, "result": result, "detail": detail})

    answer = str(doc.get("answer") or "")
    push(
        "answer_present",
        "answer is non-empty",
        "pass" if answer.strip() else "fail",
        "" if answer.strip() else "empty answer text",
    )

    verdicts = []
    for index, citation in enumerate(citations):
        url = citation.get("url") if isinstance(citation, dict) else None
        url = url if isinstance(url, str) else ""
        resolves = url in pages
        quote = citation.get("quote") if isinstance(citation, dict) else ""
        literal_ok = resolves and _norm(quote) != "" and _norm(quote) in _norm(pages[url])
        overlap = 1.0 if literal_ok else 0.0
        quote_ok = literal_ok
        if not quote_ok and resolves and mode == "fuzzy" and _norm(quote) != "":
            overlap = _token_overlap(quote, pages[url])
            quote_ok = overlap >= min_overlap
        verdicts.append(
            {
                "index": index,
                "url": url,
                "resolves": resolves,
                "quote_ok": quote_ok,
                "overlap": overlap,
            }
        )
    unresolved = [v for v in verdicts if not v["resolves"]]
    if not citations:
        citations_result = "pass" if not claims else "fail"
    else:
        citations_result = "pass" if not unresolved else "fail"
    push(
        "citations_resolve",
        "every citation url was fetched",
        citations_result,
        "" if not unresolved else f"unfetched: {', '.join(v['url'] for v in unresolved)}",
    )

    bad_quotes = [v for v in verdicts if not v["quote_ok"]]
    worst_overlap = max([v["overlap"] for v in bad_quotes] + [1.0])
    gate_name = (
        "every citation quote overlaps its page"
        if mode == "fuzzy"
        else "every citation quote matches its page literally"
    )
    if bad_quotes:
        bad_detail = (
            f"mismatch at citation index {', '.join(str(v['index']) for v in bad_quotes)}"
            f" (mode={mode}, overlap={worst_overlap:.2f})"
        )
    else:
        bad_detail = ""
    push(
        "quotes_literal",
        gate_name,
        "pass" if not bad_quotes else "fail",
        bad_detail,
    )

    uncovered = []
    for claim in claims:
        raw_urls = claim.get("citation_urls") if isinstance(claim, dict) else []
        urls = [u for u in _as_list(raw_urls) if isinstance(u, str) and u]
        claim_id = claim.get("claim_id") if isinstance(claim, dict) else None
        if not urls or any(u not in pages for u in urls):
            uncovered.append(str(claim_id or "?"))
    push(
        "claims_covered",
        "every claim points at fetched evidence",
        "pass" if not uncovered else "fail",
        "" if not uncovered else f"uncovered: {', '.join(uncovered)}",
    )

    invented = sorted(u for u in _answer_urls(answer) if u not in pages)
    push(
        "no_invented_urls",
        "answer links only fetched pages",
        "pass" if not invented else "fail",
        "" if not invented else f"invented: {', '.join(invented)}",
    )

    passed = sum(1 for gate in gates if gate["result"] == "pass")
    return {
        "format": "handoffkit.browser.model_answer_judgment",
        "format_version": 1,
        "question_id": doc.get("question_id") or "",
        "model": doc.get("model") or "",
        "mode": mode,
        "gates": gates,
        "score": (passed / len(gates)) if gates else 0,
        "verdict": "pass" if passed == len(gates) else "fail",
        "notice": "Deterministic judgment only; no claim beyond the fetched pages.",
    }
